### PR TITLE
 Enhancement/Encryption and Decryption 

### LINKED ISSUE #20: Add encryption and decryption for user password

### PR DESCRIPTION
**Pull request by Rendy752**
bcrypt hashed password result
image

### CODE REVIEW

File: src/modules/user/services/user.subservice.ts [hunk 1]
```diff
@@ -28,11 +29,17 @@ class UserSubservice extends BaseMongoService<UserDto> {
...
     const user = await this.findOne({ email: data.email });
+
     if (!user) {
       return this.throwError("User not found", StatusNotFound);
     }
 
-    if (user.password !== data.password) {
+    if (!data.password) {
+      return this.throwError("Password is required", StatusBadRequest);
+    }
```
Comment (fanesz) [inline src/modules/user/services/user.subservice.ts]:
```suggestion
```
handled by http layer (yup validator)

File: src/modules/user/services/user.service.ts [hunk 2]
```diff
@@ -50,6 +51,9 @@ class UserService extends BaseMongoService<UserDto> {
       return this.throwError("User already exists", StatusConflict);
     }
 
+    if (data.password) {
+      data.password = await bcrypt.hash(data.password, 10);
+    }
```
Comment (fanesz) [inline src/modules/user/services/user.service.ts]:
```suggestion
    data.password = await bcrypt.hash(data.password, 10);
```
unnecessary if check, handled by http layer (yup validator)

**Review by fanesz — CHANGES_REQUESTED** [review]

File: src/modules/user/services/user.service.ts (hunk 2 above)
Comment (Rendy752) [reply src/modules/user/services/user.service.ts]: i ll update to this `data.password = await bcrypt.hash(data.password as string, 10);` cause in yup the type is string | undefined

Comment (fanesz) [reply src/modules/user/services/user.service.ts]: `const isValidPassword = await bcrypt.compare(data.password!, user.password);`
@Rendy752

File: src/modules/user/services/user.subservice.ts [hunk 3]
```diff
@@ -28,11 +29,17 @@ class UserSubservice extends BaseMongoService<UserDto> {
...
   ): Promise<UserDto | ServiceError> {
     const user = await this.findOne({ email: data.email });
+
     if (!user) {
       return this.throwError("User not found", StatusNotFound);
     }
 
-    if (user.password !== data.password) {
+    const isValidPassword = await bcrypt.compare(
+      data.password as string,
```
Comment (fanesz) [inline src/modules/user/services/user.subservice.ts]:
```suggestion
      data.password!,
```

**Comment by Rendy752** [comment]
ay ay done

**Review by fanesz — APPROVED** [review]

**Pull request merged**